pkg/controller/zerocapacity/zero_capacity.go

```diff
@@ -0,0 +1,452 @@
+	err = c.Watch(&source.Kind{Type: reconciler.Type()}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	return c.Watch(&source.Kind{Type: &corev1.Pod{}}, &handler.EnqueueRequestForOwner{
+		IsController: true,
+		OwnerType:    reconciler.Type(),
+	})
+}
+
+func (z *Controller) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reconciler := z.Reconciler
+	resource := reflect.TypeOf(reconciler.Type()).Name()
+	namespace := request.Namespace
+
+	reqLogger := z.Log.WithValues("Request.Namespace", namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling " + resource)
+	defer reqLogger.Info("----- End Reconciling " + resource)
+
+	instance, err := reconciler.ResourceInstance(request.NamespacedName, z)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, fmt.Errorf("Unable to fetch %s CR '%s': %w", resource, request.Name, err)
+	}
+
+	phase := instance.Phase()
+	switch phase {
+	case "":
+		return z.initializeResources(request, instance)
+	case ZeroSucceeded, ZeroFailed:
+		return z.cleanupResources(request)
+	}
+
+	infinispan := &v1.Infinispan{}
+	clusterName := instance.Spec().Cluster
+	clusterObjKey := client.ObjectKey{
+		Namespace: namespace,
+		Name:      clusterName,
+	}
+	if err := z.Get(context.Background(), clusterObjKey, infinispan); err != nil {
+		if errors.IsNotFound(err) {
+			return reconcile.Result{}, fmt.Errorf("CR '%s' not found: %w", clusterName, err)
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, fmt.Errorf("Unable to fetch CR '%s': %w", clusterName, err)
+	}
+
+	user := consts.DefaultOperatorUser
+	pass, err := users.PasswordFromSecret(user, infinispan.GetSecretName(), namespace, z.Kube)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	httpConfig := ispnclient.HttpConfig{
+		Username:  user,
+		Password:  pass,
+		Namespace: namespace,
+		Protocol:  strings.ToLower(string(infinispan.GetEndpointScheme())),
+	}
+	httpClient := curl.New(httpConfig, z.Kube)
+
+	if phase == ZeroPending {
+		return z.execute(httpClient, request, instance)
+	}
+	// Phase must be ZeroRunning, so wait for execution to complete
+	return z.waitForExecutionToComplete(httpClient, request, instance)
+}
+
+func (z *Controller) initializeResources(request reconcile.Request, instance Resource) (reconcile.Result, error) {
+	ctx := context.Background()
+	name := request.Name
+	namespace := request.Namespace
+	spec := instance.Spec()
+	statefulset := &appsv1.StatefulSet{}
+	clusterKey := k8client.ObjectKey{
+		Namespace: namespace,
+		Name:      spec.Cluster,
+	}
+	if err := z.Client.Get(ctx, clusterKey, statefulset); err != nil {
```

Comment:
Not sure that we should validate for StatefulSet rather that Infinispan CR status.

Reply:
That's a good point. I need to get the StatefulSet PodSpec eventually, but I should validate the CR status first.